subprojects/core-api/src/main/java/org/gradle/api/artifacts/transform/TransformAction.java

```diff
@@ -31,10 +31,13 @@
  *
  * <p>A property annotated with {@link InputArtifactDependencies} will receive the <em>dependencies</em> of its input artifact.
  *
+ * @param <T> Parameter type for the transform action. Should be {@link Void} if the action does not have parameters.
  * @since 5.3
  */
+@SuppressWarnings("unused")
 @Incubating
-public interface TransformAction {
+public interface TransformAction<T> {
```

Comment:
We should have the getter for the parameters object declared on this interface, if there's a type parameter for the parameter object on it. We'd also get rid of `@TransformParameters` as we'd no longer need it.

I would also use a separate interface for transforms with parameters and one without. Otherwise we haven't really made anything simpler, we've just moved the surprises around. So, we might have `TransformAction` with no type parameter or getter to represent transforms in general, and have a `TransformActionWithParameters<T>` (or whatever) that extends this for those transforms that happen to need a parameters object in addition to the parameters provided by Gradle. The separation allows us to evolve the APIs more easily and doesn't bake 'every transform has parameters' into the API.